app/src/components/fineTunes/FineTuneContentTabs/General/InferenceCodeTabs/InferenceCodeTabs.tsx

```diff
@@ -0,0 +1,131 @@
+    model="openpipe:{{FINE_TUNE_MODEL_SLUG}}",
+    messages={{TEMPLATED_MESSAGES}},
+    tool_choice={{TEMPLATED_TOOL_CHOICE}},
+    tools={{TEMPLATED_TOOLS}},
+    openpipe={
+        "tags": {
+            "prompt_id": "counting",
+            "any_key": "any_value"
+        }
+    },
+)
+
+print(completion.choices[0].message)`,
+  },
+  {
+    title: "JavaScript",
+    language: "javascript",
+    code: `import OpenAI from "openpipe/openai";
+
+const client = new OpenAI({
+    openpipe: {
+        apiKey: "{{TEMPLATED_OPENPIPE_API_KEY}}",
+    }
+});
+
+const completion = await client.chat.completions.create({
+    model: "openpipe:{{FINE_TUNE_MODEL_SLUG}}",
+    messages: {{TEMPLATED_MESSAGES}},
+    tool_choice: {{TEMPLATED_TOOL_CHOICE}},
+    tools: {{TEMPLATED_TOOLS}},
+});
+
+console.log(completion?.choices[0]?.message);
+`,
+  },
+];
+
+const templateTabs = (
+  apiKey: string,
+  modelSlug: string,
+  entry?: RouterOutputs["datasetEntries"]["listTrainingEntries"]["entries"][number]["datasetEntry"],
+): CodeTab[] => {
+  let formattedMessages = "";
+
+  if (entry?.messages) {
+    const messages = entry.messages || [
+      {
+        role: "system",
+        content: "Count to 10",
+      },
+    ];
+    // add a tab to all but the first line
+    formattedMessages = JSON.stringify(messages, null, 4);
+  }
+
+  const formattedToolChoice = JSON.stringify(entry?.tool_choice ?? "auto", null, 4);
+
+  const formattedTools = JSON.stringify(entry?.tools ?? []);
+
+  return baseTabs.map((tab) => {
+    const templatedCode = tab.code;
+
+    const tabFormattedMessages =
+      tab.language === "bash"
+        ? formattedMessages.replace(/'/g, `'"'"'`)
+        : formattedMessages.replace(/\n/g, "\n    ");
+    const tabFormattedToolChoice =
+      tab.language === "bash"
+        ? formattedToolChoice.replace(/'/g, `'"'"'`)
+        : formattedToolChoice.replace(/\n/g, "\n    ");
+    const tabFormattedTools =
+      tab.language === "bash" ? formattedTools.replace(/'/g, `'"'"'`) : formattedTools;
+
+    return {
+      ...tab,
+      code: templatedCode
+        .replace("{{TEMPLATED_OPENPIPE_API_KEY}}", apiKey)
+        .replace("{{FINE_TUNE_MODEL_SLUG}}", modelSlug)
+        .replace("{{TEMPLATED_TOOL_CHOICE}}", tabFormattedToolChoice.replace(/`/g, "\\`"))
+        .replace("{{TEMPLATED_TOOLS}}", tabFormattedTools.replace(/`/g, "\\`"))
+        .replace("{{TEMPLATED_MESSAGES}}", tabFormattedMessages.replace(/`/g, "\\`")),
+    };
+  });
+};
+
+const InferenceCodeTabs = () => {
+  const project = useSelectedProject().data;
+  const fineTune = useFineTune().data;
+  const datasetEntries = useTrainingEntries().data;
+
+  const templatedTabs = useMemo(() => {
+    const entry = datasetEntries?.entries[0];
+
+    return templateTabs(project?.openpipeApiKey || "", fineTune?.slug || "", entry?.datasetEntry);
+  }, [project?.openpipeApiKey, fineTune?.slug, datasetEntries?.entries]);
+
+  return <CopiableCodeTabs tabs={templatedTabs} />;
```

Comment:
I'm a little confused by the code organization here. Why is this a separate nested component? They seem very strongly linked.

Reply:
Because there is already a ton of logic in the top-level component, and I like to limit component size and complexity.

Reply:
Encapsulating logic makes a ton of sense when it's reused, and can make sense even without reuse if it presents a clean and intuitive interface so the caller doesn't have to understand the inner workings of the callee. However, in this case realistically if you're working in these tabs you're going to have to understand both components anyway, so splitting them up doesn't really gain us much. In fact it's the opposite -- I found myself having to bounce back and forth between these files several times to figure out which pieces are defined where.

Reply:
It makes sense that you had to flit between both of these files to understand their combined functionality when first reading through the code. I think we should continue to separate the logic for determining what to display from the logic for displaying it to cut down on the amount of logic in each file - otherwise we'll end up with an instance of that brand of enormous monolithic component that makes development so unpleasant.